File: JavaScript/ES2015/Behavioral/strategy.js
class MeowSound {
    makeSound() {
        return 'Meow meow meow!';
    }
}

class PurrSound {
    makeSound() {
        return 'Purrr... Purrrrr...';
    }
}

class Cat {
    constructor(name) {
        this._name = name;
        this._soundBehavior = null;
    }

    performSound() {
        console.log(`${this._name}: ${this._soundBehavior.makeSound()}`);
    }

    setSoundBehavior(sb) {
        this._soundBehavior = sb;
    }
}

class Siamese extends Cat {
    constructor(name) {
        super(name);
        this._soundBehavior = new MeowSound();
    }
}

class MaineCoon extends Cat {
    constructor(name) {
        super(name);
        this._soundBehavior = new PurrSound();
    }
}

let siamese = new Siamese('Simba'),
    maineCoon = new MaineCoon('Oscar');

siamese.performSound();
maineCoon.performSound();
siamese.setSoundBehavior(new PurrSound());
siamese.performSound();

File: behavioral/strategy/strategy.cs
class StrategyPattern
{
    interface SoundBehavior
    {
        string MakeSound();
    }

    class MeowSound : SoundBehavior
    {
        public string MakeSound()
        {
            return "Meow meow meow!";
        }
    }

    class PurrSound : SoundBehavior
    {
        public string MakeSound()
        {
            return "Purrr... Purrrrr...";
        }
    }

    abstract class Cat
    {
        protected SoundBehavior soundBehavior { get; set; }
        protected string name { get; }

        public Cat(string name)
        {
            this.name = name;
        }

        public void PerformSound()
        {
            System.Console.WriteLine("{0}: {1}", name, soundBehavior.MakeSound());
        }

        public void SetSoundBehavior(SoundBehavior ab)
        {
            soundBehavior = ab;
        }
    }

    class Siamese : Cat
    {
        public Siamese(string name)
            : base(name)
        {
            soundBehavior = new MeowSound();
        }
    }

    class MaineCoon : Cat
    {
        public MaineCoon(string name)
            : base(name)
        {
            soundBehavior = new PurrSound();
        }
    }

    static void Main(string[] args)
    {
        Cat simba = new Siamese("Simba");
        Cat oscar = new MaineCoon("Oscar");

        simba.PerformSound();
        oscar.PerformSound();
        simba.SetSoundBehavior(new PurrSound());
        simba.PerformSound();
    }
}
